File: src/main/java/br/com/blz/testjava/mapper/impl/ProductMapper.java
package br.com.blz.testjava.mapper.impl;

import br.com.blz.testjava.dto.InventoryDto;
import br.com.blz.testjava.dto.ProductDto;
import br.com.blz.testjava.entity.Inventory;
import br.com.blz.testjava.entity.Product;
import br.com.blz.testjava.entity.Warehouse;
import br.com.blz.testjava.mapper.AbstractMapper;

public class ProductMapper implements AbstractMapper<Product, ProductDto>{

    public ProductMapper() {
        super();
    }

    @Override
    public ProductDto fromEntityToDto(Product entity) {

        InventoryDto inventoryDto = new InventoryDto();
        inventoryDto.setWarehouses(entity.getInventory().getWarehouses());
        long _quantity = entity.getInventory().getWarehouses()
                .stream().mapToLong(Warehouse::getQuantity).sum();
        inventoryDto.setQuantity(_quantity);

        boolean _isMarkable = _quantity>0l;

        return new ProductDto(entity.getSku(), entity.getName(), inventoryDto, _isMarkable);
    }

    @Override
    public Product fromDtoToEntity(ProductDto dto) {

        Inventory inventory = new Inventory();
        inventory.setWarehouses(dto.getInventory().getWarehouses());
        inventory.setQuantity(dto.getInventory().getQuantity());

        return new Product(dto.getSku(), dto.getName(), inventory, dto.isMarketable());
    }

}

File: src/main/kotlin/br/com/blz/testjava/product/mapper/ProductMapper.kt
package br.com.blz.testjava.product.mapper

import br.com.blz.testjava.inventory.mapper.InventoryMapper
import br.com.blz.testjava.product.dto.ProductInDTO
import br.com.blz.testjava.product.dto.ProductOutDTO
import br.com.blz.testjava.product.model.Product

class ProductMapper {
  companion object {

    fun toEntity(productInDTO: ProductInDTO): Product {
      return Product(
        productInDTO.sku,
        productInDTO.name,
        InventoryMapper.toEntity(productInDTO.inventory)
      )
    }

    fun toOutDto(product: Product): ProductOutDTO {
      return ProductOutDTO(
        product.sku,
        product.name,
        InventoryMapper.toOutDto(product.inventory),
        product.isMarketable
      )
    }
  }
}
